src/app/utils/numberFormatter.ts

```diff
@@ -0,0 +1,42 @@
+import { BigNumber } from 'bignumber.js'
+
+export type NumberFormattingParameters = Partial<BigNumber.Format> & {
+  // Additional features which are not natively supported by BigNumber
+
+  decimalPlaces?: number
+  maximumFractionDigits?: number
+  roundingMode?: BigNumber.RoundingMode
+  unit?: string
+}
+
+export const formatNumber = (
+  inputNumber: number | string | BigNumber.Instance | undefined,
+  format: NumberFormattingParameters = {},
+): string | undefined => {
+  if (inputNumber === undefined) return
+  const { decimalPlaces, maximumFractionDigits, roundingMode, unit, ...formatting } = format
+  let number =
+    typeof inputNumber === 'number'
+      ? BigNumber(inputNumber.toString(2), 2) // This is required to keep all precision
```

Comment:
Read [this](https://github.com/MikeMcl/bignumber.js/blob/d87c6606f905668445b1f5782f94adf8c02c4618/README.md?plain=1#L97) and [this](https://github.com/MikeMcl/bignumber.js/blob/d87c6606f905668445b1f5782f94adf8c02c4618/README.md?plain=1#L115) together.

> If the limited precision of Number values is not well understood, it is recommended to create BigNumbers from String values rather than Number values to avoid a potential loss of precision.

> When creating a BigNumber from a Number, note that a BigNumber is created from a Number's decimal `toString()` value not from its underlying binary value. If the latter is required, then pass the Number's `toString(2)` value and specify base 2.

Reply:
So, to summarize: according to the docs, in order to get all the data from a Number, with no loss of precision, we need to create a binary string representation. Otherwise it will just create a string (with whatever precision) and that will be the input of BigNumber.

Reply:
okay it loses precision *differently* 
```ts
BigNumber((10000000000/3).toString(2), 2).toFixed()
'3333333333.33333349227905273438'

BigNumber(10000000000/3).toFixed()
'3333333333.3333335'
```